### ISSUE TITLE
Can't create custom deletion hooks with Composition Functions

### ISSUE THREAD

**Issue by ytsarev** [crossplane/crossplane#5092]
### What happened?

I want to execute custom logic during the XR deletion with the Composition Function.

Apparently, the Composition code is not getting executed at all after the XR is getting `deletionTimestamp`

As @phisco pointed out that the composite reconciler is [exiting](https://github.com/crossplane/crossplane/blob/d346a67573e5cd111a7631bdb0f86ed9c0914204/internal/controller/apiextensions/composite/reconciler.go#L506) and not running Function pipeline after observing the `deletionTimestamp` on the Composite.

Looks like there is no way to hook in the custom Function logic seamlessly.

At the same time, the [Composition Functions documentation ](https://docs.crossplane.io/latest/concepts/composition-functions/#write-a-composition-function) claims that the Functions suppose to handle custom deletion logic.

I trust that it is a valid important use case for Composition Functions. Some example use cases include:

* when deleting the AWS Organizational Unit, move it to another OU instead of the actual deletion
* when deleting Azure PostgresqlSever, perform complex custom backup and only then delete

### How can we reproduce it?

Please check https://github.com/ytsarev/function-pause-when-ready/ with the representative logic and examples.

The logic around XR deletion timestamp https://github.com/ytsarev/function-pause-when-ready/blob/main/fn.go#L46

https://github.com/ytsarev/function-pause-when-ready/blob/main/fn_test.go#L153 - the associated passing unit test

The e2e test will fail on the deletion stage:

```
k apply -f example/definition.yaml
k apply -f example/composition.yaml
k apply -f example/xr.yaml
```

Wait until the first `bucket` reaches the readiness state and will be automatically paused by the function.

```
k get bucket
NAME                READY   SYNCED   EXTERNAL-NAME   AGE
any-xr-vj6fn        True    False    any-xr-vj6fn    13s
```

```
k describe bucket any-xr-vj6fn
....
Events:
  Type    Reason                   Age   From                                            Message
  ----    ------                   ----  ----                                            -------
  Normal  CreatedExternalResource  28s   managed/s3.aws.upbound.io/v1beta1, kind=bucket  Successfully requested creation of external resource
  Normal  ReconciliationPaused     16s   managed/s3.aws.upbound.io/v1beta1, kind=bucket  Reconciliation is paused either through the `spec.managementPolicies` or the pause annotation
```

```
 k get bucket any-xr-vj6fn -oyaml | grep paused
    crossplane.io/paused: "true"
```

Notice that  `crossplane.io/paused: "true"` annotation was automatically set by the function.

Now delete the XR 

```
 k delete -f example/xr.yaml
anyxr.test.fn.upbound.io "any-xr" deleted
```

and observe that the `bucket` is stuck on deletion because `crossplane.io/paused: "true"` is never getting set to `"false"` 

Effectively the [condition](https://github.com/ytsarev/function-pause-when-ready/blob/main/fn.go#L50) that is supposed to react to populated `deletionTimestamp` is never executed.

### What environment did it happen in?
Crossplane version: 1.14

**Comment by phisco** [comment]
Yes, this is correct, functions can not do this today, https://github.com/crossplane/crossplane/issues/4729 should cover this capability.

Regarding the docs, it is indeed wrong, currently functions are not called on deletion.

**Comment by negz** [comment]
I tend to agree with @phisco that these use cases seem better suited to hypothetical "operations functions" (https://github.com/crossplane/crossplane/issues/4729) rather than _composition_ functions. So we should update the docs to clarify how functions behave for a deleted XR.

Composition functions exist to tell Crossplane what composed resources to create, update, and delete when an XR is created, updated, or deleted. The "delete" part is pretty much implied; if the XR is deleted all of its composed resources should be deleted. I'd argue that functions should not have side-effects on external systems (i.e. creating things in external systems) and thus should not have any external clean-up to do.

Assuming we did want to run the function pipeline when the XR was deleted, how would that work? What would the function be expected to return? Would it signal that composed resources should be deleted by returning an empty set of desired composed resources? If so, every function would need logic to determine whether the XR was deleted (e.g. by looking for a deletion timestamp) so that it could return an empty set of desired resources.

**Comment by ytsarev** [comment]
/fresh we still have no solution for the custom deletion logic

**Comment by ytsarev** [comment]
/fresh we still have no solution for the custom deletion logic

**Comment by negz** [comment]
> when deleting Azure PostgresqlSever, perform complex custom backup and only then delete

I've been working on the Operation (day two ops) design at https://github.com/crossplane/crossplane/pull/6557. We realized they won't work well for this specific use case - see https://github.com/crossplane/crossplane/pull/6557#discussion_r2158544810 for context.

You'll be able to use Operations to hook into deletes of arbitrary resources, so you can trigger stuff to happen on a delete. What you can't do is _block_ the delete (e.g. to run a backup first).

If we need to block deletes until some work is done probably making XRs run the function pipeline on delete is the only way to do that. This way the XR doesn't trigger deletion of the composed resource at all until it's e.g. backed it up.

This has some serious gotchas to work through:

* What is a function expected to return as desired state when the XR has been deleted?
  * Does it keep returning the composed resources until it's safe to deleted them?
  * If so, every function needs updating to handle the delete case... 
* Do we need to make foreground/background delete configurable e.g. in XR schema (😢), like we did with claims?

**Comment by bobh66** [comment]
> What you can't do is block the delete (e.g. to run a backup first).

Breadcrumbs to a possibly related comment [here](https://github.com/crossplane/crossplane/pull/6557#discussion_r2167825121) - would require changes to the MR reconciler but might be worth considering?

> Do we need to make foreground/background delete configurable e.g. in XR schema (😢), like we did with claims?

Or maybe just always use `foreground` as described [here](https://github.com/crossplane/crossplane/issues/6288)?  I've tried to think of reasons not to and so far I haven't come up with any.

**Comment by kaessert** [comment]
Hitting the same and i completely disagree with operation functions solving this. Operation functions would be late and i see this as simply as submitting a paramter to the function request. Even if existing functions don't do anything with this parameter i don't see a problem, they will just output the same desired resource one more time before the XR is deleted but giving functions which might retain a state at least a chance to manage that state including deletions. 
Deletion of the XR could block in case there are function errors and would give us some sort of a stable base to handle this cases.
